### PR TITLE
Message details view-models

### PR DESCRIPTION
**Pull request by BarAgam**
Created message details view-models and instantiated the different kinds at GGApplication for future use

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/teamagam/GimelGimel/pull/60" target="_blank">ReviewNinja

File: app/src/main/java/com/teamagam/gimelgimel/app/GGApplication.java [hunk 1]
```diff
@@ -27,6 +30,9 @@
     private MessagesReadStatusModel mMessagesReadStatusModel;
     private SelectedMessageModel mSelectedMessageModel;
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/GGApplication.java:31]: consider create class for holding all models and view models

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/ImageMessageDetailViewModel.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Image message details view-model
+ */
+public class ImageMessageDetailViewModel extends MessageDetailViewModel {
+    public ImageMessageDetailViewModel(
+            SelectedMessageModel selectedMessageModel) {
+        super(selectedMessageModel);
+    }
+
+    public String getImageUrl() {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/ImageMessageDetailViewModel.java:19]: you missed here getSource(). !fix

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/LatLongMessageDetailViewModel.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * LatLong message view-model
+ */
+public class LatLongMessageDetailViewModel extends MessageDetailViewModel {
+    public LatLongMessageDetailViewModel(
+            SelectedMessageModel selectedMessageModel) {
+        super(selectedMessageModel);
+    }
+
+    public PointGeometry getPointGeometry() {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/LatLongMessageDetailViewModel.java:16]: consider change it's name to getLocation?

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessageDetailViewModel.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import com.teamagam.gimelgimel.app.common.DataChangedObservable;
+import com.teamagam.gimelgimel.app.model.ViewsModels.Message;
+import com.teamagam.gimelgimel.app.model.entities.messages.SelectedMessageModel;
+
+/**
+ * Shared functionality of detail message view-model
+ * Verifies selected message is of appropriate type before each method call.
+ * Notifies observers on data changes only when selected message is of appropriate type.
+ */
+abstract class MessageDetailViewModel extends SelectedMessageViewModel implements DataChangedObservable {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessageDetailViewModel.java]: why it is not public?
I would like to hold reference to MessageDetailViewModel so i would not need to update the title every time.

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/SelectedMessageViewModel.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        return mSelectedMessageModel.getSelected().getSenderId();
+    }
+
+    public Date getDate() {
+        validateSelectedMessage();
+        return mSelectedMessageModel.getSelected().getCreatedAt();
+    }
+
+    protected void validateSelectedMessage() {
+        if (!mSelectedMessageModel.isAnySelected()) {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/SelectedMessageViewModel.java:41]: isAnySelected is a useful method also for the views

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/SelectedMessageViewModel.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            throw new NoSelectedMessageException();
+        }
+    }
+
+    protected abstract boolean shouldNotifyObservers();
+
+    public static class NoSelectedMessageException extends RuntimeException {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/SelectedMessageViewModel.java]: if you want to control the flow with this exception make also the upper class public !resolve

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/LatLongMessageDetailViewModel.java (hunk 3 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/LatLongMessageDetailViewModel.java:16]: but it returns a point geometry

File: app/src/main/java/com/teamagam/gimelgimel/app/GGApplication.java (hunk 1 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/GGApplication.java:31]: The next task adds another field. 
I'll do it there

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/ImageMessageDetailViewModel.java (hunk 2 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/ImageMessageDetailViewModel.java:19]: !fixed

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessageDetailViewModel.java (hunk 4 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessageDetailViewModel.java]: Fine, you're right.
There, I said it.
Happy now ?

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/SelectedMessageViewModel.java (hunk 6 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/SelectedMessageViewModel.java]: change it to have the default access modifier. !resolved

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/SelectedMessageViewModel.java (hunk 5 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/SelectedMessageViewModel.java:41]: The container viewmodel wil expose such a method

**Pull request merged**